Source/objects.cpp

```diff
@@ -587,39 +591,41 @@ void AddObjTraps()
 		rndv = 25;
 	for (int j = 0; j < MAXDUNY; j++) {
 		for (int i = 0; i < MAXDUNX; i++) {
-			if (dObject[i][j] <= 0 || GenerateRnd(100) >= rndv)
+			Object *triggerObject = ObjectAtPosition({ i, j }, false);
+			if (triggerObject == nullptr || GenerateRnd(100) >= rndv)
 				continue;
 
-			int8_t oi = dObject[i][j] - 1;
-			if (!AllObjects[Objects[oi]._otype].oTrapFlag)
+			if (!AllObjects[triggerObject->_otype].oTrapFlag)
 				continue;
 
+			Object *trapObject = nullptr;
 			if (GenerateRnd(2) == 0) {
 				int xp = i - 1;
-				while (IsTileNotSolid({ xp, j })) // BUGFIX: check if xp >= 0
+				while (IsTileNotSolid({ xp, j }))
 					xp--;
 
 				if (!CanPlaceWallTrap(xp, j) || i - xp <= 1)
 					continue;
 
 				AddObject(OBJ_TRAPL, { xp, j });
-				int8_t oiTrap = dObject[xp][j] - 1;
-				Objects[oiTrap]._oVar1 = i;
-				Objects[oiTrap]._oVar2 = j;
-				Objects[oi]._oTrapFlag = true;
+				trapObject = ObjectAtPosition({ xp, j });
 			} else {
 				int yp = j - 1;
-				while (IsTileNotSolid({ i, yp })) // BUGFIX: check if yp >= 0
+				while (IsTileNotSolid({ i, yp }))
 					yp--;
 
 				if (!CanPlaceWallTrap(i, yp) || j - yp <= 1)
 					continue;
 
 				AddObject(OBJ_TRAPR, { i, yp });
-				int8_t oiTrap = dObject[i][yp] - 1;
-				Objects[oiTrap]._oVar1 = i;
-				Objects[oiTrap]._oVar2 = j;
-				Objects[oi]._oTrapFlag = true;
+				trapObject = ObjectAtPosition({ i, yp });

```

Comment:
not required, but returning the object from `AddObject` would probably make sens

Reply:
`AddObject` needs work, it currently partially initialises an `Object` instance while also placing it on the map. I'd rather see it only place the object (similar to `container::push_back`) and have the caller fully initialise the object instead of having to do it in two stages like this.

Reply:
Hmm, having said that this could be considered an analogue to `container::emplace_back` which returns a reference. I'll add that to the list for a future PR :)